tests/solidity/test_close_short.py

```diff
@@ -0,0 +1,364 @@
+import elfpy.pricing_models.hyperdrive as hyperdrive_pm
+import elfpy.time as time
+import elfpy.types as types
+
+
+# pylint: disable=too-many-arguments
+# pylint: disable=duplicate-code
+
+
+class TestCloseShort(unittest.TestCase):
+    """
+    Test opening a opening and closing a short in hyperdrive
+    3 failure cases: zero amount, invalid amount, invalid timestamp
+    6 success cases:
+        - close immediately, with regular, and small amounts
+        - redeem at maturity, with zero interest, and with negative interest (skipped)
+        - close halfway thru term, with zero, and negative interest (both skipped)
+    """
+
+    contribution: float = 500_000_000
+    target_apr: float = 0.05
+    term_length: int = 365
+    alice: agent.Agent
+    bob: agent.Agent
+    celine: agent.Agent
+    hyperdrive: hyperdrive_market.Market
+
+    def setUp(self):
+        """Set up agent, pricing model, & market for the subsequent tests.
+        This function is run before each test method.
+        """
+        self.alice = agent.Agent(wallet_address=0, budget=self.contribution)
+        self.bob = agent.Agent(wallet_address=1, budget=self.contribution)
+        block_time = time.BlockTime()
+        pricing_model = hyperdrive_pm.HyperdrivePricingModel()
+        market_state = hyperdrive_market.MarketState(
+            trade_fee_percent=0.0,
+            redemption_fee_percent=0.0,
+        )
+        self.hyperdrive = hyperdrive_market.Market(
+            pricing_model=pricing_model,
+            market_state=market_state,
+            position_duration=time.StretchedTime(
+                days=365, time_stretch=pricing_model.calc_time_stretch(self.target_apr), normalizing_constant=365
+            ),
+            block_time=block_time,
+        )
+        _, wallet_deltas = self.hyperdrive.initialize(self.alice.wallet.address, self.contribution, 0.05)
+        self.alice.wallet.update(wallet_deltas)
+
+    def verify_close_short(
+        self,
+        example_agent: agent.Agent,
+        market_state_before: hyperdrive_market.MarketState,
+        agent_base_paid: float,
+        agent_base_proceeds: float,
+        bond_amount: float,
+        maturity_time: float,
+    ):
+        """Close a short then make sure the market state is correct"""
+        # verify that all of Bob's bonds were burned
+        self.assertFalse(
+            example_agent.wallet.shorts
+        )  # In solidity we check that the balance is zero, but here we delete the entry if it is zero
+        # verify that the bond reserves were updated according to flat+curve
+        # the adjustment should be equal to timeRemaining * bondAmount
+        time_remaining = 0
+        if maturity_time > self.hyperdrive.block_time.time:
+            time_remaining = self.hyperdrive.position_duration.normalized_time * (
+                maturity_time - self.hyperdrive.block_time.time
+            )
+        self.assertLess(  # user gets less than what they put in
```

Comment:
Since this is an extra test (pulled in from other tests), I think it makes sense to put it first. The primary goal here is for Alex to be able to quickly review all of the python tests and verify parity against the solidity tests. So we want to implement the functions & asserts in the same order with approximately the same names.

Reply:
this is equivalent to these two in solidity

CloseLongTest.t.sol:
```
// Verify that Bob didn't receive more base than he put in.
        assertLe(baseProceeds, basePaid);
```

CloseShortTest.t.sol:
```
// Verify that Bob doesn't end up with more base than he started with.
        assertGe(basePaid, baseProceeds);
```

Reply:
Yes, I'm just suggesting you list it first in this function